provisioning/transport/mqtt/src/mqtt.ts

```diff
@@ -4,12 +4,413 @@
+  /* Codes_SRS_NODE_PROVISIONING_MQTT_18_050: [ When connecting, `Mqtt` shall set `host` in the base `TransportConfig` object to the `provisioningDeviceHost`.] */
+  /* Codes_SRS_NODE_PROVISIONING_MQTT_18_035: [ When connecting, `Mqtt` shall set `clientId` in the base `registrationRequest` object to the registrationId.] */
+  /* Codes_SRS_NODE_PROVISIONING_MQTT_18_036: [ When connecting, `Mqtt` shall set the `clean` flag in the base `TransportConfig` object to true.] */
+  /* Codes_SRS_NODE_PROVISIONING_MQTT_18_038: [ When connecting, `Mqtt` shall set the `username` in the base `TransportConfig` object to '<idScope>/registrations/<registrationId>/api-version=<apiVersion>&clientVersion=<UrlEncode<userAgent>>'.] */
+  let baseConfig: MqttBase.TransportConfig = {
+      host: request.provisioningHost,
+      deviceId: request.registrationId,
+      clean: true,
+      x509: this._auth,
+      username: request.idScope + '/registrations/' + request.registrationId + '/api-version=' + ProvisioningDeviceConstants.apiVersion + '&ClientVersion=' + encodeURIComponent(ProvisioningDeviceConstants.userAgent)
+    };
+
+    /* Codes_SRS_NODE_PROVISIONING_MQTT_18_039: [ If a uri is specified in the request object, `Mqtt` shall set it in the base `TransportConfig` object.] */
+    let uri: string = (<any>request).uri;
+    if (uri && uri.indexOf('wss://') === 0) {
+      (<any>baseConfig).uri = uri;
+    }
+
+    /* Codes_SRS_NODE_PROVISIONING_MQTT_18_040: [ When connecting, `Mqtt` shall call `_mqttBase.connect`.] */
+    this._mqttBase.connect(baseConfig, (err) => {
+      if (err) {
+        /* Codes_SRS_NODE_PROVISIONING_MQTT_18_041: [ If an error is returned from `_mqttBase.connect`, `Mqtt` shell wrap it in a `TransportSpecificError` object and pass it to the caller using `callback`.] */
+        debug('connect error: ' + err.toString());
+        callback(new errors.TransportSpecificError('connect error', err));
+      } else {
+        /* Codes_SRS_NODE_PROVISIONING_MQTT_18_042: [ After connecting the transport, `Mqtt` will subscribe to '$dps/registrations/res/#'  by calling `_mqttBase.subscribe`.] */
+        this._mqttBase.subscribe(responseTopic, { qos: 1 }, (err) => {
+          if (err) {
+            /* Codes_SRS_NODE_PROVISIONING_MQTT_18_043: [ If an error is returned from _mqttBase.subscribe, `Mqtt` shell wrap it in a `TransportSpecificError` object and pass it to the caller using `callback`.] */
+            debug('subscribe error: ' + err.toString());
+            callback(new errors.TransportSpecificError('subscribe error',err));
+          } else {
+            debug('connected and subscribed successfully');
+            callback();
+          }
+        });
+      }
+    });
+  }
+
+  private _disconnect(callback: (err?: Error) => void): void {
+
+    /* Codes_SRS_NODE_PROVISIONING_MQTT_18_044: [ When Disconnecting, `Mqtt` shall call _`mqttBase.unsubscribe`.] */
+    this._mqttBase.unsubscribe(responseTopic, (unsubscribeError) => {
```

Comment:
>unsubscribe [](start = 19, length = 11)

shouldn't we call this only if subscription happened and not every time we cancel ?